### ISSUE TITLE
IHostedService StopAsync and OnStopping/OnStopped events not called with IIS preload enabled

### ISSUE THREAD

**Issue by lpn-ch** [dotnet/aspnetcore#28089]
In summary, if I

- Install the "Application Initialization module" in IIS
- Set "StartMode: AlwaysRunning" for the app pool
- Set "Preload Enabled: True" for the website

The IHostedService StopAsync method and OnStopping/OnStopped events are not called when either recycling the app pool or restarting IIS (iisreset). Just disabling "Preload Enabled" fixes the issue.

Not sure this is an IIS or .Net Core issue.

Here is the output of dotnet --info

Host (useful for support):
  Version: 3.1.9
  Commit:  774fc3d6a9

.NET Core SDKs installed:
  No SDKs were found.

.NET Core runtimes installed:
  Microsoft.AspNetCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.9 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

IIS on Windows Server 2019 1809, but same issue on Windows 10 1909.

Code sample:

public class Program
{
	public static void Main(string[] args)
	{
		CreateHostBuilder(args).Build().Run();
	}

	public static IHostBuilder CreateHostBuilder(string[] args) =>
		Host.CreateDefaultBuilder(args)
			.ConfigureWebHostDefaults(webBuilder =>
			{
				webBuilder.UseStartup<Startup>();
			})
			.ConfigureServices(services =>
			{
				services.AddHostedService<LifetimeEventsHostedService>();
			});
}

internal class LifetimeEventsHostedService : IHostedService
{
    private readonly ILogger _logger;
    private readonly IHostApplicationLifetime _appLifetime;

    public LifetimeEventsHostedService(
        ILogger<LifetimeEventsHostedService> logger, 
        IHostApplicationLifetime appLifetime)
    {
        _logger = logger;
        _appLifetime = appLifetime;
    }

    public Task StartAsync(CancellationToken cancellationToken)
    {
        _appLifetime.ApplicationStarted.Register(OnStarted);
        _appLifetime.ApplicationStopping.Register(OnStopping);
        _appLifetime.ApplicationStopped.Register(OnStopped);

        return Task.CompletedTask;
    }

    public Task StopAsync(CancellationToken cancellationToken)
    {
        _logger.LogInformation("StopAsync has been called.");

        return Task.CompletedTask;
    }

    private void OnStarted()
    {
        _logger.LogInformation("OnStarted has been called.");
    }

    private void OnStopping()
    {
        _logger.LogInformation("OnStopping has been called.");
    }

    private void OnStopped()
    {
        _logger.LogInformation("OnStopped has been called.");
    }
}

StartAsync and OnStarted are always called. It seems the process is killed by IIS instead of cleanly shutting down when preload is enabled.

**Comment by BrennanConroy** [comment]
Could you try this with 5.0?

Also, can you verify that this isn't just an issue with the logs not being flushed by debugging the app on close?

**Comment by lpn-ch** [comment]
> Could you try this with 5.0?
> 
> Also, can you verify that this isn't just an issue with the logs not being flushed by debugging the app on close?

Forgive me for the late reply, I was very busy last week.

After doing more tests, I have more details concerning the issue, but I will first answer your questions.

- .Net 5.0 does not solve the issue.
- It is not a log not flushed issue, I replaced with File.AppendAllText() which closes the file after write and the issue is still there. Also, there would be no reason to log in one case and not the other.

But the issue of the preload not calling stop events on recycle/iisreset is actually a side effect of another issue. If the website does not receive any requests, the stop events are never called on recycle/iisreset. If the website does receive one request, the stop events start getting called on every recycle until an iisreset is done which will be the last to call stop events. This is why the issue is not exhibited with preload disabled since a request has to be done to activate the website, which in turn solves the issue.

Here is the sequence to reproduce:

- Configure preload to work on the app pool and wesite.
- App pool recycling or iisreset will not call stop events.
- Make a single request on the website.
- App pool recycling or iisreset will call stop events.
- App pool recycling will call stop events until iisreset is executed.

Hope this helps.

**Comment by jkotalik** [comment]
>  If the website does not receive any requests, the stop events are never called on recycle/iisreset

In the case where preload is disabled, I believe this expected as the start events haven't been called either.

And to confirm, when you have preload enabled, you see the start events occur without sending a request, but on recycle you don't see the stop events called. Only after sending a request to the server, do you see the stop events called.

**Comment by lpn-ch** [comment]
Author feedback is requested but I do not understand why since the issue has been acknowledged.

Maybe jkotalik was waiting for me to confirm this:

"And to confirm, when you have preload enabled, you see the start events occur without sending a request, but on recycle you don't see the stop events called. Only after sending a request to the server, do you see the stop events called."

This statement is correct.

**Comment by MoonStorm** [comment]
This is a pretty serious issue and needs to be escalated. The IHostedService.StopAsync doesn't seem to have been tested under IIS and is simply not working.  The background workers are being killed without any means to clean up gracefully.

**Comment by jkotalik** [comment]
We currently have this in next sprint planing, which means we will try to investigate this issue over the coming months.

**Comment by notemac** [comment]
**Having the same problem!**

Environment: .NET 6, ASP.NET Core Web API project, IIS (10.0.19041.320), Windows 21H2 (OS Build 19044.1586).

IIS Configuration: 
* ApplicationInitializationModule installed
* preloadEnabled = True, 
* startMode = AlwaysRunning
* idleTimeout = 0
* time = 0

Web.config:
```XML
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <location path="." inheritInChildApplications="false">
    <system.webServer>
      <handlers>
        <add name="aspNetCore" path="*" verb="*" modules="AspNetCoreModuleV2" resourceType="Unspecified" />
      </handlers>
      <aspNetCore processPath=".\Server1.exe" stdoutLogEnabled="false" stdoutLogFile=".\logs\stdout" hostingModel="inprocess" />
    </system.webServer>
  </location>
    <system.webServer>
        <applicationInitialization doAppInitAfterRestart="false">
            <add initializationPage="/main" />
        </applicationInitialization>
    </system.webServer>
</configuration>
```

## Test case
1) Start **Application Pool** and 4 files created:
* 23828 2022-04-08 16_38_14.6463605 ctor
* 23828 2022-04-08 16_38_14.6531225 StartAsync
* 23828 2022-04-08 16_38_14.6576097 OnStarted
* 23828 2022-04-08 16_38_14.7171710 PreLoad  

2) Start **Site** related to this Application Pool and 6 files created:
* **23828 2022-04-08 16_42_11.8621622 StopAsync**
* **23828 2022-04-08 16_42_11.8627045 OnStopped**
* 21460 2022-04-08 16_42_12.2586683 ctor
* 21460 2022-04-08 16_42_12.2653510 StartAsync
* 21460 2022-04-08 16_42_12.2695714 OnStarted
* 21460 2022-04-08 16_42_12.3222581 PreLoad

3) Stop **Application Pool** and 0 files created:

4) Again start **Application Pool** and 4 files created:
* 23944 2022-04-08 16_49_23.2675547 ctor
* 23944 2022-04-08 16_49_23.2743613 StartAsync
* 23944 2022-04-08 16_49_23.2784152 OnStarted
* 23944 2022-04-08 16_49_23.3308574 PreLoad

5) Send request to "/main" and 1 file created:
* 23944 2022-04-08 16_50_53.0517791 PreLoad

6) Stop **Application Pool** and, oops, 2 files created:
* **23944 2022-04-08 16_51_44.9061052 StopAsync**
* **23944 2022-04-08 16_51_44.9065086 OnStopped**

Program.cs:
```c#
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddControllers();
builder.Services.AddHostedService<HostApplicationLifetimeEventsHostedService>();
var app = builder.Build();
app.UseAuthorization();
app.MapControllers();
app.Run();
```
HostApplicationLifetimeEventsHostedService.cs:
```c#
public class HostApplicationLifetimeEventsHostedService : IHostedService
{
    private readonly IHostApplicationLifetime _applicationLifetime;
    public HostApplicationLifetimeEventsHostedService(IHostApplicationLifetime applicationLifetime)
    {
        _applicationLifetime = applicationLifetime;
        using var _ = System.IO.File.Create($"{Environment.ProcessId} {DateTime.Now:yyyy-MM-dd HH_mm_ss.fffffff} ctor");
    }
    public Task StartAsync(CancellationToken cancellationToken)
    {
        using var _ = System.IO.File.Create($"{Environment.ProcessId} {DateTime.Now:yyyy-MM-dd HH_mm_ss.fffffff} StartAsync");
        _applicationLifetime.ApplicationStopped.Register(OnStopped);
        _applicationLifetime.ApplicationStarted.Register(OnStarted);
        return Task.CompletedTask;
    }

    private void OnStopped()
    {
        using var _ = System.IO.File.Create($"{Environment.ProcessId} {DateTime.Now:yyyy-MM-dd HH_mm_ss.fffffff} OnStopped");
    }

    private void OnStarted()
    {
        using var _ = System.IO.File.Create($"{Environment.ProcessId} {DateTime.Now:yyyy-MM-dd HH_mm_ss.fffffff} OnStarted");
    }

    public Task StopAsync(CancellationToken cancellationToken)
    {
        using var _ = System.IO.File.Create($"{Environment.ProcessId} {DateTime.Now:yyyy-MM-dd HH_mm_ss.fffffff} StopAsync");
        return Task.CompletedTask;
    }
}
```
Controller.cs
```c#
    [Route("[controller]")]
    [ApiController]
    public class MainController : ControllerBase
    {
        [HttpGet]
        public IActionResult Main(CancellationToken cancellationToken)
        {
            using var _ = System.IO.File.Create($"{Environment.ProcessId} {DateTime.Now:yyyy-MM-dd HH_mm_ss.fffffff} PreLoad");
            return Ok();
        }
    }
```

**Comment by CoraxGit** [comment]
Please work on that issue. Waiting for .NET 8 is no choice here.

Such problems sometimes make me doubt whether .NET >= 5 is really the right choice for productive systems

**Comment by joey-powell** [comment]
I'm on the latest release of .NET 7, and I'm still seeing this problem. Do you guys know when it will be fixed? Thanks

**Comment by jkappel71** [comment]
It has been 3 years since this problem was reported, and supposedly added to the next sprint backlog.  I am still seeing the problem today in .NET 7 as well.  We could really use a fix!

**Comment by BrennanConroy** [comment]
I believe this will coincidentally be fixed by https://github.com/dotnet/aspnetcore/pull/52807.
The root cause seems to be that if a request is never received by the app, then when the app is shut down `GL_STOP_LISTENING` is never triggered which is what we were using to trigger the managed server to shut down.

**Comment by amcasey** [comment]
Hopefully, this is already fixed but, otherwise, it's not something we're actively investigating.

**Comment by janpahnen** [comment]
Since the fix seems to be backported to .NET 8 with opt-in in PR #55288 but it was not mentioned how to opt-in. How is it activcated?

**Comment by martincostello** [comment]
Looking at [the diff](https://github.com/dotnet/aspnetcore/pull/55288/files#diff-07d3eb7c79ae26fa79fe46c8517d9cdab7a7aff209147df39530dd10dbd2c9e6) it appears to be either:

1. Add a `shutdownDelay="<non-negative integer>"` attribute to the `aspNetCore` element in `Web.config`
2. Set the `ANCM_shutdownDelay=<non-negative integer>` environment variable

**Comment by BrennanConroy** [comment]
Marking as completed as it should be fixed and no one has said otherwise for almost 2 years.
